### ISSUE TITLE
Tokenize and Parse PHP source code

### ISSUE THREAD

**Issue by abertschi** [abertschi/graalphp#12]
We compare two approaches to parse PHP source code. Adaption of the parser 
implemented in JPHP and use of an existing grammer for ANTRL to generate a new Parser.

JPHP
```
(- is contra, + is pro)

+ PHP 7.1+ compliant
+ Apache License 2.0
- Tokenizer and Parser written from scratch
  - a lot of boilerplate code
  - tokenizer and syntax analyser code: > 10k LOC
  + Parser creates instances of Token for all language constructs
  - Token instances do not provide visitor pattern for tree transformations
- dependency to org.ow2.asm (reflections)
- Poorly documented code
+ Active community, 1.6k stars on github
- 100k LOC in total (core: 38k, runtime 51k, extensions, package, debugger...)
+ implements PHP standard library
```

How parsing is implemented:
- Parsing is orchestrated by SyntaxAnalyser.java
- There is a Token class for each language construct
- Generator<Token> classes perform parsing and store parsed properties (modifier, isFinal, name, etc.) in Token classes
- Generator classes glue token tree together
  (methods contain alot of 'instanceof' and usually several hundres
  lines long, hand written)
- all generators are part of the SyntaxAnalyser
- other generators use SyntaxAnalyser to get result of other generators 
  (SyntaxAnalyser#generator(ConstGenerator.class).getToken(current))

In a next step, SyntaxAnalyser is used in Compiler Classes. Compiler
classes create Entity classes;

- Entity classes represent classes, methods, etc with newly gained evaluation results during compilation
- Compiler classes do transformation to bytecode
- Entity classes can potentially be used as AST POJOs but they are
  strictly connected to compiler classes which makes them hard to
  separate
- Tokenizer tree could be reused in graalphp but they are verbose and many LOC
- if SyntaxAnalyser is reused in graalphp, semantic analysis, module support, and error handling can potentially be reused
  - though PHP is not strongly typed, and Scalar Type Declarations are not checked in SynaxAnalyser

What to adapt if JPHP parser is used
- figure out how to translate JPHP tree to graalphp AST tree
- rewrite Token classes to 'accept' a visitor, and return all their children
- or find other way to iterate through token tree (many instanceof selections)

ANTLR
```
(- is contra, + is pro)

+ BSD 3-clause license
+ grammar for PHP 7.4 available
+ Generates Visitor or listener classes for tree traversal
- no provided semantic analysis. needs to be implemented where needed
- no PHP specific error messages, needs to be implemented where needed
+ vertical development allows expeditious progress, we don't need to implement the whole AST visitor
```

How parsing is implemented
- Generate ANTRL Parser and Lexer based on grammar
- Write AST visitor. Iterate over ANTRL tokens and create graaphp-node classes
- full traversal over antrl tree returns a graalphp rootnode

The ANTRL approach seems more straightforward and introduces less code to maintain.
The token parsing code is generated and the use of the visitor pattern allows for elegant
translation to a graal AST tree. ANTRL is used in other graal projects, too.

**Comment by abertschi** [comment]
JPHP SyntaxAnalyser: 
- https://github.com/jphp-group/jphp/blob/master/jphp-core/src/org/develnext/jphp/core/syntax/SyntaxAnalyzer.java


ANTRL PHP Grammar:
- https://github.com/antlr/grammars-v4/tree/master/php

**Comment by abertschi** [comment]
Benchmarking showed that the ANTRL parser is slow compared to the original PHP parser and JPHP (up to 100x slower).

```
~/.../main/resources >>> find . | xargs wc                        
wc: .: Is a directory
      0       0       0 .
      3       5      73 ./test2.php
     61     269    1973 ./test.php
wc: ./phpoffice: Is a directory
      0       0       0 ./phpoffice
   7948   26260  285899 ./phpoffice/Xls.php
     21     169    1079 ./phpoffice/LICENSE
wc: ./benchmarkgame: Is a directory
      0       0       0 ./benchmarkgame
    143     387    3136 ./benchmarkgame/fasta.php-3.php
     16      64     470 ./benchmarkgame/binarytrees.php-2.php
     99     216    2871 ./benchmarkgame/nbody.php-3.php
     64     171    1095 ./benchmarkgame/spectralnorm.php-2.php
    123     319    2874 ./benchmarkgame/fasta.php-2.php
     51     259    1783 ./benchmarkgame/fannkuchredux.php-1.php
     61     164    1361 ./benchmarkgame/binarytrees.php-1.php

# average parsing speed in ms after warm up
BenchmarkParser.antrlParser                                          phpoffice/Xls.php    avgt       5  1590.528 ± 122.401   ms/op
BenchmarkParser.antrlParser                        benchmarkgame/binarytrees.php-1.php    avgt       5    13.981 ±   0.204   ms/op
BenchmarkParser.antrlParser                      benchmarkgame/fannkuchredux.php-1.php    avgt       5    54.455 ±   0.376   ms/op
BenchmarkParser.antrlParser                                                   test.php    avgt       5    36.484 ±   0.614   ms/op
BenchmarkParser.jphpParser                                           phpoffice/Xls.php    avgt       5    15.030 ±   0.270   ms/op
BenchmarkParser.jphpParser                         benchmarkgame/binarytrees.php-1.php    avgt       5     0.388 ±   0.017   ms/op
BenchmarkParser.jphpParser                       benchmarkgame/fannkuchredux.php-1.php    avgt       5     0.464 ±   0.011   ms/op
BenchmarkParser.jphpParse                                                     test.php    avgt       5     0.452 ±   0.011   ms/op

# cold start speed
BenchmarkParser.antrlParser                                          phpoffice/Xls.php      ss          3627.084             ms/op
BenchmarkParser.antrlParser                        benchmarkgame/binarytrees.php-1.php      ss           263.001             ms/op
BenchmarkParser.antrlParser                      benchmarkgame/fannkuchredux.php-1.php      ss           289.005             ms/op
BenchmarkParser.antrlParser                                                   test.php      ss           451.512             ms/op
BenchmarkParser.jphpParser                                           phpoffice/Xls.php      ss           331.019             ms/op
BenchmarkParser.jphpParser                         benchmarkgame/binarytrees.php-1.php      ss           145.861             ms/op
BenchmarkParser.jphpParser                       benchmarkgame/fannkuchredux.php-1.php      ss           156.936             ms/op
BenchmarkParser.jphpParser                                                    test.php      ss           141.496             ms/op
```

Above listing shows JMH benchmarks on selected sample files:

- avgt is average parsing execution after JVM warmup
- ss is cold start execution
- we see that antrl parser is 100x slower than JPHP;
- Xls.php is 8000 lines long, execution takes 1590 ms with ANTRL, 15 ms with JPHP (after warm up)
- binarytree.php-1.php is 61 lines, execution takes 13 ms with ANTRL, 0.4ms with JPHP


Native PHP parsing speed of Xls.php is in the range of 10ms
```
~/.../resources/phpoffice >>> perf stat -r 10 -B php -l Xls.php
No syntax errors detected in Xls.php
No syntax errors detected in Xls.php
No syntax errors detected in Xls.php
No syntax errors detected in Xls.php
No syntax errors detected in Xls.php
No syntax errors detected in Xls.php
No syntax errors detected in Xls.php
No syntax errors detected in Xls.php
No syntax errors detected in Xls.php
No syntax errors detected in Xls.php

 Performance counter stats for 'php -l Xls.php' (10 runs):

             10.50 msec task-clock:u              #    0.965 CPUs utilized            ( +-  0.51% )
                 0      context-switches:u        #    0.000 K/sec                  
                 0      cpu-migrations:u          #    0.000 K/sec                  
             1,613      page-faults:u             #    0.154 M/sec                    ( +-  0.04% )
        28,924,264      cycles:u                  #    2.755 GHz                      ( +-  0.25% )
        47,378,960      instructions:u            #    1.64  insn per cycle           ( +-  0.00% )
         9,736,313      branches:u                #  927.426 M/sec                    ( +-  0.00% )
           234,643      branch-misses:u           #    2.41% of all branches          ( +-  0.13% )

         0.0108791 +- 0.0000619 seconds time elapsed  ( +-  0.57% )
```         

ANTRL parser is too slow and thus not an option to adopt for graalphp.

- options:
  - test ANTRL with native image and see if performance much better (it is not)
  - stick to ANTRL, try to optimize grammar for speedup
    - unclear if optimization gives much speedup
    - there are many nestings in antrl grammar
    - html and js rules implemented as well, removal may give speedup
  - switch to JPHP, find a way to convert tree
    - implies that we need to main sourcecode (10k loc)
    - code is ugly, hard(er) to maintain
    - code uses reflection (causes troubles with graal native-installer, I did not manage to compile it)
  - write better ANTRL grammar from scratch
    - may take several weeks

**Comment by abertschi** [comment]
I stumbled upon the Eclipse PHP Development Tools. Eclipse provides a PHP flavor of its IDE. It comes with a Java parser for PHP. https://www.eclipse.org/pdt/

Benchmarking of their parser showed promising results. Similar speeds to JPHP and native PHP.
Challenges:
- Parser is tightly integrated in the eclipse tool suite
- Dependencies to dltk, UI code and eclipse runtime

I extracted their parser into a standalone project. It will be used for graalphp.
